lib/vsc/filesystem/oceanstor.py

```diff
@@ -151,11 +172,229 @@ def __init__(self, url, username=None, password=None):
+
+        if not isinstance(sp_names, list):
+            target_storage_pools = [sp_names]
+        else:
+            target_storage_pools = sp_names
+
+        try:
+            sp_select = [storage_pools[sp]['storagePoolName'] for sp in target_storage_pools]
+        except KeyError as err:
+            sp_miss = err.args[0]
+            sp_avail = ", ".join(storage_pools)
+            errmsg = "Storage pool '%s' not found in OceanStor. Use any of: %s" % (sp_miss, sp_avail)
+            self.log.raiseException(errmsg, KeyError)
+
+        # Convert names to IDs
+        if byid:
+            sp_select = [storage_pools[sp]['storagePoolId'] for sp in sp_select]
+
+        return sp_select
+
+    def list_filesystems(self, device=None, update=False):
+        """
+        List all filesystems
+
+        Set self.filesystems to a convenient dict structure of the returned dict
+        where the key is the filesystem name, the value is a dict with keys:
+        - atime_update_mode
+        - dentry_table_type
+        - dir_split_bitwidth
+        - dir_split_policy
+        - id
+        - is_show_snap_dir
+        - name
+        - qos_policy_id
+        - rdc
+        - record_id
+        - root_split_bitwidth
+        - running_status
+        - storage_pool_id
+        """
+
+        # Filter by requested devices (storage pools in OceanStor)
+        # Support special case 'all' for downstream compatibility
+        if device is None or device == 'all':
+            storage_pools = self.list_storage_pools(update=update)
+            device = list(storage_pools.keys())
+
+        sp_ids = self.select_storage_pools(device, byid=True)
+        filter_sp_ids = [{'storage_pool_id': str(sp_id)} for sp_id in sp_ids]
+        filter_sp_ids_json = json.dumps(filter_sp_ids, separators=self.json_sep)
+        self.log.debug("Filtering filesystems in storage pools with IDs: %s", ', '.join(str(i) for i in sp_ids))
+
+        if not update and self.filesystems:
+            # Use cached filesystem data
+            filesystems = {fs['name']: fs for fs in self.filesystems.values() if fs['storage_pool_id'] in sp_ids}
+            dbg_prefix = "(cached) "
+        else:
+            # Request filesystem data
+            _, response = self.session.file_service.file_systems.get(filter=filter_sp_ids_json)
+            filesystems = {fs['name']: fs for fs in response['data']}
+            dbg_prefix = ""
+
+            self.filesystems = filesystems
+
+        self.log.debug(dbg_prefix + "Filesystems in OceanStor: %s", ", ".join(filesystems))
+
+        return filesystems
+
+    def select_filesystems(self, filesystemnames, devices=None, byid=False):
+        """
+        Return list of existing filesytem with the provided filesystem names
+        Restrict list to filesystems found in given storage pools names
+
+        @type filesystemnames: list of filesystem names (if string: 1 filesystem; if None: all known filesystems)
```

Comment:
How is `None` done in the code below?

Reply:
I forgot to update the description. This method only handles lists or a string. `None` is only possible with `list_filesets()` for compatibility with `GpfsOperations.list_filesets()`. Updated description in https://github.com/vub-hpc/vsc-filesystem-oceanstor/pull/2/commits/ddd1a4487dd34ed7b2c0b3d2317ab7c4c2958aca